Adds execution timer stats to the range query

davkal opened the pull request:
API consumers should be able to get insight into the query run times.
The UI currently measures total roundtrip times. This PR allows for more
fine grained metrics to be exposed.

* adds new timer for total execution time (queue + eval)

* expose new timer, queue timer, and eval timer in stats field of the
 range query response:
```json
{
  "status": "success",
  "data": {
    "resultType": "matrix",
    "result": [],
    "stats": {
      "execQueueTimeNs": 4683,
      "execTotalTimeNs": 2086587,
      "totalEvalTimeNs": 2077851
    }
  }
}
```

* stats field is optional, only set when query parameter `stats` is not
empty

Try it via
```sh
curl 'http://localhost:9090/api/v1/query_range?query=up&start=1486480279&end=1486483879&step=14000&stats=true'
```

beorn7 commented:
@juliusv @fabxc I think any of you knows this code better than me. Could you have a look?

juliusv commented on web/api/v1/api.go:
```diff
@@ -236,9 +244,23 @@ func (api *API) queryRange(r *http.Request) (interface{}, *apiError) {
 		return nil, &apiError{errorExec, res.Err}
 	}
+
+	// optional stats field in response if parameter "stats" is set
```
Start comments capitalized, end with period please.

juliusv commented on web/api/v1/api.go:
```diff
@@ -236,9 +244,23 @@ func (api *API) queryRange(r *http.Request) (interface{}, *apiError) {
+
+	// optional stats field in response if parameter "stats" is set
+	var qs *queryStats
+	if s := r.FormValue("stats"); s != "" {
```
This can just be `if r.FormValue("stats") != ""`

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -148,9 +149,16 @@ func (api *API) Register(r *route.Router) {
 }
 
+type queryStats struct {
+	ExecQueueTime time.Duration `json:"execQueueTimeNs"`
```
Seconds are the convention in Prometheus

davkal replied:
Seconds even for internal measurements? I guess if the timers are in low nanos, the numbers won't matter much, because that's fast. However, as an API consumer I do care about times in the millies. And I would like the stats timers to be of the same unit. Currently the API response would expose the `time.duration` as measured by Go, which happens to be in `ns`.

- [ ] Keep nanos for precision, or
- [ ] Use millies for API consumer relevance, or
- [ ] Use seconds, with decimals (how many?) for consistency

juliusv commented on util/stats/timer.go:
```diff
@@ -45,6 +45,11 @@ func (t *Timer) ElapsedTime() time.Duration {
 	return time.Since(t.start)
 }
 
+// Duration returns the duration value of the timer (in ns).
```
It returns it as a `time.Duration`, which has no unit per se, so omit the `(in ns)`.

juliusv commented on web/api/v1/api.go:
```diff
@@ -148,9 +149,16 @@ func (api *API) Register(r *route.Router) {
 	r.Get("/alertmanagers", instr("alertmanagers", api.alertmanagers))
 }
 
+type queryStats struct {
```
Maybe rather than defining this struct here it would be better to organize all possible timers as members of a struct (maybe as pointers so some counters can be `nil` depending on kind of query) in query_stats.go instead of a list of consts (https://github.com/prometheus/prometheus/blob/master/util/stats/query_stats.go#L20-L28). And then return the entire struct (JSON-ized), not just some of the timers. This'd require some refactoring of the timer code.

grobie replied:
@juliusv Hmm, but this moves the API specification outside of the API package. A change to the util package naming will also change the API now.

fabxc commented:
@alexsomesan ties in to your work on actually instrumenting those stats. Right?
Could you take a look?

alexsomesan commented:
Yes, I have some pending changes that expose these as metrics.
We should also include these in the exposed set.

davkal commented:
I addressed the comments, PTAL:

* moved query stats json generation to query_stats.go
* use seconds for all query timers
* expose all timers available

A new example response looks like:
```json
{
  "status": "success",
  "data": {
    "resultTypes": "matrix",
    "result": [],
    "stats": {
      "totalEvalTime": 0.002703116,
      "resultSortTime": 0.000392294,
      "queryPreparationTime": 0.001849997,
      "innerEvalTime": 0.000156043,
      "resultAppendTime": 2.16e-07,
      "execQueueTime": 0.000622995,
      "execTotalTime": 0.003337607
    }
  }
}
```

beorn7 commented:
@juliusv Could you pick this up?

juliusv commented on util/stats/query_stats.go:
```diff
@@ -42,7 +43,47 @@ func (s QueryTiming) String() string {
 	case ExecQueueTime:
 		return "Exec queue wait time"
+	case ExecTotalTime:
+		return "Exec queue total time"
```
"Exec queue" -> "Exec"?

juliusv commented:
Besides my nit, I think this is fine for now, but I see future potential around refactoring the timer code more so that we have less duplication of all the field names.

grobie commented on util/stats/query_stats.go:
```diff
@@ -25,6 +25,7 @@ const (
 	InnerEvalTime
 	ResultAppendTime
 	ExecQueueTime
+	ExecTotalTime
```
Hm, we have now both `TotalEvalTime` and `ExecTotalTime`, which doesn't seem to follow a consistent naming pattern.

Gouthamve replied:
Yep, can you make it `EvalTotalTime`?

grobie commented on util/stats/query_stats.go:
```diff
@@ -42,7 +43,47 @@ func (s QueryTiming) String() string {
+
+// MakeQueryStats makes a QueryStats struct with all QueryTimings found in the
+// given TimerGroup.
+func MakeQueryStats(tg *TimerGroup) *QueryStats {
```
Why did you chose the prefix `Make` over `New` here? Do you have a link to a guide line?

juliusv commented:
> Given this will be part of the v1 API, we won't be able to change it once merged.

What exact timers are returned will depend on the storage and query engine internals, so that should be allowed to change without breaking the API... as in, we should not guarantee in the API documentation that any specific stats are returned.

Gouthamve commented:
@davkal This looks pretty useful, could you pick this up and rebase? If not, I am happy to carry this PR.

davkal commented:
@Gouthamve rebase done, PTAL.
I also added the optional stats to the `query` endpoint.

davkal commented:
Hmrpf, a SD test failed on Travis.

Gouthamve commented:
:+1:

grobie commented on util/stats/stats_test.go:
```diff
@@ -0,0 +1,55 @@
+
+	var qs *QueryStats
+	qs = NewQueryStats(tg)
+	actual, _ := json.Marshal(qs)
```
Please handle errors, for example with a t.Fatal.

davkal commented:
Great, SD test failed again.

Gouthamve commented:
Retriggered them :) Will be fixed soon.
